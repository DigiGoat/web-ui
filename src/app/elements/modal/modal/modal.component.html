<div #modal class="modal fade" id="exampleModal" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
  <div class="modal-dialog modal-dialog-centered modal-dialog-scrollable modal-lg">
    <div class="modal-content">
      <div class="modal-header color-scheme-tertiary">
        <h1 class="modal-title fs-5" id="staticBackdropLabel" test-id="searchParam">{{searchParam}}</h1>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body color-scheme-quaternary card">
        <app-modal-goat-card *ngIf="goat" [goat]="goat" test-id="goat-card"></app-modal-goat-card>
        <app-modal-not-found *ngIf="!goat" [searchParam]="searchParam" test-id="not-found"></app-modal-not-found>
      </div>
      <div class="modal-footer color-scheme-tertiary">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>
