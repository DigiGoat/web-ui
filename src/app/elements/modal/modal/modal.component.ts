import { Component, Input, ViewChild, type AfterViewInit, type ElementRef, type OnDestroy, type OnInit } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import type { Goat } from '../../../services/goat/goat.service';
import { PlatformService } from '../../../services/platform/platform.service';

@Component({
  selector: 'app-modal-goat',
  templateUrl: './modal.component.html',
  styleUrl: './modal.component.scss'
})
export class ModalComponent implements OnDestroy, AfterViewInit, OnInit {
  prerender = false;
  @Input() goat?: Partial<Goat>;
  @Input({ required: true }) searchParam!: string;

  constructor(public router: Router, public route: ActivatedRoute, private platformService: PlatformService) { }

  ngOnInit() {
    this.prerender = this.platformService.isServer;
  }
  ngOnDestroy() {
    this.close();
  }

  @ViewChild('modal') modalElement!: ElementRef<HTMLDivElement>;
  private modal?: bootstrap.Modal;
  ngAfterViewInit() {
    if (!this.prerender) {
      this.modal = bootstrap.Modal.getOrCreateInstance(this.modalElement.nativeElement);
      this.open();
      this.modalElement.nativeElement.addEventListener('hidden.bs.modal', () => {
        this.router.navigate(['../'], { relativeTo: this.route });
      });
    }
  }

  open() {
    this.modal?.show();
  }
  close() {
    this.modal?.hide();
    this.modal?.dispose();
  }
}
