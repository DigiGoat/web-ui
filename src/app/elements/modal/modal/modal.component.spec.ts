import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ActivatedRoute, type Router } from '@angular/router';
import type { PlatformService } from '../../../services/platform/platform.service';
import { ModalComponent } from './modal.component';

jest.mock('@angular/router');
const bootstrap = { Modal: { getOrCreateInstance: jest.fn().mockReturnValue({ dispose: jest.fn(), show: jest.fn(), hide: jest.fn() }) } };
describe('ModalComponent', () => {
  let component: ModalComponent;
  let fixture: ComponentFixture<ModalComponent>;
  let html: HTMLElement;

  beforeEach(async () => {
    (window as unknown as { bootstrap: object; }).bootstrap = bootstrap;
    await TestBed.configureTestingModule({
      declarations: [ModalComponent],
      providers: [ActivatedRoute]
    })
      .compileComponents();

    fixture = TestBed.createComponent(ModalComponent);
    component = fixture.componentInstance;
    html = fixture.nativeElement;
    component['router'] = { navigate: jest.fn() } as unknown as Router;
    component.searchParam = 'TEST_PARAM';
  });
  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });
  describe('In a browser', () => {
    beforeEach(() => {
      component['platformService'] = { isServer: false, isBrowser: true, isBot: false } as PlatformService;
      fixture.detectChanges();
    });
    it('should initialize the Modal', () => {
      expect(bootstrap.Modal.getOrCreateInstance).toHaveBeenCalledTimes(1);
    });
    it('should open the modal', () => {
      expect(component['modal']?.show).toHaveBeenCalledTimes(1);
    });
    it('should navigate backwards when the modal closes', () => {
      component.modalElement.nativeElement.dispatchEvent(new Event('hidden.bs.modal'));
      expect(component['router'].navigate).toHaveBeenCalledTimes(1);
      expect(component['router'].navigate).toHaveBeenCalledWith(['../'], { relativeTo: component['route'] });
    });
    it('should close the modal when destroyed', () => {
      fixture.destroy();
      expect(component['modal']?.hide).toHaveBeenCalledTimes(1);
      expect(component['modal']?.dispose).toHaveBeenCalledTimes(1);
    });
  });
  describe('On the Server', () => {
    beforeEach(() => {
      component['platformService'] = { isServer: true, isBrowser: false, isBot: false } as PlatformService;
      fixture.detectChanges();
    });
    it('should not initialize the Modal', () => {
      expect(bootstrap.Modal.getOrCreateInstance).not.toHaveBeenCalled();
    });
  });
  describe('With a Goat', () => {
    beforeEach(() => {
      component.goat = { name: 'Billy' };
      fixture.detectChanges();
    });
    it('should set the header to the searchParam', () => {
      const element = html.querySelector('[test-id=searchParam]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_PARAM');
    });
    it('should display the goat card', () => {
      const element = html.querySelector('[test-id=goat-card]');
      expect(element).toBeTruthy();
    });
    it('should not display the not-found card', () => {
      const element = html.querySelector('[test-id=not-found]');
      expect(element).toBeFalsy();
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
  describe('Without a Goat', () => {
    beforeEach(() => {
      fixture.detectChanges();
    });
    it('should set the header to the searchParam', () => {
      const element = html.querySelector('[test-id=searchParam]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_PARAM');
    });
    it('should display not the goat card', () => {
      const element = html.querySelector('[test-id=goat-card]');
      expect(element).toBeFalsy();
    });
    it('should display the not-found card', () => {
      const element = html.querySelector('[test-id=not-found]');
      expect(element).toBeTruthy();
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
});
