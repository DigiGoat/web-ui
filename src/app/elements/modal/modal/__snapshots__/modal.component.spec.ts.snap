// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalComponent With a Goat should match the snapshot 1`] = `
<app-modal-goat
  goat={[Function Object]}
  modal={[Function Object]}
  modalElement={[Function _ElementRef]}
  platformService={[Function PlatformService]}
  prerender="false"
  route={[Function ActivatedRoute]}
  router={[Function Object]}
  searchParam={[Function String]}
>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalLabel"
    class="modal fade"
    id="exampleModal"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered modal-dialog-scrollable modal-lg"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header color-scheme-tertiary"
        >
          <h1
            class="modal-title fs-5"
            id="staticBackdropLabel"
            test-id="searchParam"
          >
            TEST_PARAM
          </h1>
          <button
            aria-label="Close"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <div
          class="modal-body color-scheme-quaternary card"
        >
          <app-modal-goat-card
            test-id="goat-card"
          />
        </div>
        <div
          class="modal-footer color-scheme-tertiary"
        >
          <button
            class="btn btn-secondary"
            data-bs-dismiss="modal"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</app-modal-goat>
`;

exports[`ModalComponent Without a Goat should match the snapshot 1`] = `
<app-modal-goat
  modal={[Function Object]}
  modalElement={[Function _ElementRef]}
  platformService={[Function PlatformService]}
  prerender="false"
  route={[Function ActivatedRoute]}
  router={[Function Object]}
  searchParam={[Function String]}
>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalLabel"
    class="modal fade"
    id="exampleModal"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-dialog-centered modal-dialog-scrollable modal-lg"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header color-scheme-tertiary"
        >
          <h1
            class="modal-title fs-5"
            id="staticBackdropLabel"
            test-id="searchParam"
          >
            TEST_PARAM
          </h1>
          <button
            aria-label="Close"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <div
          class="modal-body color-scheme-quaternary card"
        >
          <app-modal-not-found
            test-id="not-found"
          />
        </div>
        <div
          class="modal-footer color-scheme-tertiary"
        >
          <button
            class="btn btn-secondary"
            data-bs-dismiss="modal"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</app-modal-goat>
`;
